runner/pkg/actions/mkvod.go

```diff
@@ -0,0 +1,142 @@
+	streamID, ok := d["streamID"].(uint64)
+	if !ok {
+		return AbortingError(fmt.Errorf("no stream id in context"))
+	}
+	streamVersion, ok := d["streamVersion"].(string)
+	if !ok {
+		return AbortingError(fmt.Errorf("no stream end in context"))
+	}
+	recordingDir, ok := d["recordingDir"].(string)
+	if !ok {
+		return AbortingError(fmt.Errorf("no recordingDir in context"))
+	}
+
+	vodDir := path.Join(config.Config.StoragePath, fmt.Sprintf("%d", streamID), streamVersion)
+	err := os.MkdirAll(vodDir, os.ModePerm)
+	if err != nil {
+		return AbortingError(fmt.Errorf("create VOD directory: %w", err))
+	}
+
+	recordingContent, err := os.ReadDir(recordingDir)
+	if err != nil {
+		return AbortingError(fmt.Errorf("read recordingDir: %w", err))
+	}
+
+	for _, entry := range recordingContent {
+		if entry.IsDir() {
+			log.Warn("found dir in recordingDir, skipping", "name", entry.Name())
+			continue
+		}
+		if entry.Name() == "playlist.m3u8" {
+			srcPlst, err := os.Open(path.Join(recordingDir, entry.Name()))
+			if err != nil {
+				return AbortingError(fmt.Errorf("open recording playlist: %w", err))
+			}
+			srcPlstB, err := io.ReadAll(srcPlst)
+			if err != nil {
+				return AbortingError(fmt.Errorf("read recording playlist: %w", err))
+			}
+			dstPlstS := vodFromEventPlst(string(srcPlstB))
+			dstPlst, err := os.Create(path.Join(vodDir, entry.Name()))
+			if err != nil {
+				return AbortingError(fmt.Errorf("create vod playlist: %w", err))
+			}
+			_, err = io.WriteString(dstPlst, dstPlstS)
+			if err != nil {
+				return fmt.Errorf("write vod to playlist: %w", err)
+			}
+			_ = dstPlst.Close()
+			continue
+		}
+		err = copyFile(path.Join(recordingDir, entry.Name()), path.Join(vodDir, entry.Name()))
+	}
+
+	vodUrl, err := url.JoinPath(config.Config.EdgeServer, fmt.Sprintf("%d", streamID), streamVersion, "playlist.m3u8")
+	if err != nil {
+		return fmt.Errorf("join vod url: %w", err)
+	}
+	notify <- &protobuf.Notification{
+		Data: &protobuf.Notification_VodReady{
+			VodReady: &protobuf.VODReadyNotification{
+				Stream:        &protobuf.StreamInfo{Id: ptr.Take(streamID)},
+				StreamVersion: ptr.Take(protobuf.StreamVersion(protobuf.StreamVersion_value[streamVersion])),
+				Url:           ptr.Take(vodUrl),
+			},
+		},
+	}
+	return nil
+}
+
+func copyFile(sourcePath, destPath string) error {
+	inputFile, err := os.Open(sourcePath)
+	if err != nil {
+		return fmt.Errorf("open source file: %w", err)
+	}
+	defer inputFile.Close()
+
+	outputFile, err := os.Create(destPath)
+	if err != nil {
+		return fmt.Errorf("open dest file: %w", err)
+	}
+	defer outputFile.Close()
+
+	_, err = io.Copy(outputFile, inputFile)
+	if err != nil {
+		return fmt.Errorf("copy to dest from source: %w", err)
+	}
+	return nil
+}
+
+// vodFromEventPlst modifies an HLS playlist from EVENT to VOD
+func vodFromEventPlst(playlist string) string {
+	var lines []string
+	hasEndlist := false
+
+	scanner := bufio.NewScanner(strings.NewReader(playlist))
```

Comment:
Okay.. so to read a file line by line one has to read it as binary, convert to string and then scan the string?

Could you not skip the first part?

Reply:
good remark, I changed vodFromEventPlst accept an io.Reader instead 👍